### PR TITLE
Base: + Sketcher: disable geometry deselection when user holds Ctrl.

### PR DESCRIPTION
**Pull request by PaddleStroke**
This is solving the deselection problem for sketcher and for general 3D view.

https://github.com/FreeCAD/FreeCAD/issues/7673
https://forum.freecadweb.org/viewtopic.php?f=8&t=73056
https://forum.freecadweb.org/viewtopic.php?f=8&t=72435

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7699">pipeline status</a> for feature branch [PR_7699](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7699). Pipeline [685489225](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/685489225) was triggered at [3b950a1](https://github.com/FreeCAD/FreeCAD/pull/7699/commits/3b950a10cd21e5eb49ee55ea1e252b5e2589f796). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7699">pipeline status</a> for feature branch [PR_7699](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7699). Pipeline [685640843](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/685640843) was triggered at [9795999](https://github.com/FreeCAD/FreeCAD/pull/7699/commits/9795999a9ce73b0fdfbbccc053ad428c661378f3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by chennes** [comment]
Thanks for the quick PR -- looks great, nice improvement to the behavior.

**Comment by PaddleStroke** [comment]
Thanks for looking into it ! :)

File: src/Gui/NavigationStyle.cpp [hunk 1]
```diff
@@ -1479,8 +1479,10 @@ SbBool NavigationStyle::processEvent(const SoEvent * const ev)
     // check for left click without selecting something
     if ((curmode == NavigationStyle::SELECTION || curmode == NavigationStyle::IDLE)
             && !processed) {
-        if (SoMouseButtonEvent::isButtonReleaseEvent(ev,SoMouseButtonEvent::BUTTON1))
-            Gui::Selection().clearSelection();
+        if (SoMouseButtonEvent::isButtonReleaseEvent(ev, SoMouseButtonEvent::BUTTON1)) {
+            if (!(QApplication::keyboardModifiers() == Qt::ControlModifier))
```
Comment (wwmayer) [inline src/Gui/NavigationStyle.cpp:1483]: @PaddleStroke I think this is the wrong check. If you additionally press another key like Shift or Alt then it still clears the selection. Btw you can also use Inventor API: `if (!ev->wasCtrlDown())`

File: src/Mod/Sketcher/Gui/ViewProviderSketch.cpp [hunk 2]
```diff
@@ -889,9 +889,10 @@ bool ViewProviderSketch::mouseButtonPressed(int Button, bool pressed, const SbVe
                     }
                     Mode = STATUS_NONE;
                     return true;
-                case STATUS_SKETCH_StartRubberBand: // a single click happened, so clear selection
+                case STATUS_SKETCH_StartRubberBand: // a single click happened, so clear selection unless user hold control.
+                    if (!(QApplication::keyboardModifiers() == Qt::ControlModifier))
```
Comment (wwmayer) [inline src/Mod/Sketcher/Gui/ViewProviderSketch.cpp:893]: @PaddleStroke This suffers from the same problem but the Inventor API can't be used here. So, the check should be: `if (!(QApplication::keyboardModifiers() & Qt::ControlModifier))`

**Comment by PaddleStroke** [comment]
Thanks for the comments.
Why using inventor API rather than the same at both places?
Do you want me to open a PR for that or do you want to commit directly to
fix it?

**Comment by wwmayer** [comment]
> Why using inventor API rather than the same at both places?

Actually, in all places when we have an SoEvent we use the methods this class provides and avoid to mix two APIs too much.

The next thing is that the signature of `mouseButtonPressed(int Button, bool pressed, const SbVec2s &cursorPos, const Gui::View3DInventorViewer *viewer)` is a design flaw which leads to the use of magic numbers (e.g. for Button) and it suppresses some extra information that might be needed here and there.
So, long story short: at some point this must be cleaned-up and an SoMouseEvent should be directly passed. Then one could consistently use the Inventor API in the implementations of the sub-classes.

> Do you want me to open a PR for that or do you want to commit directly to fix it?

If you agree that the code should be adjusted I can do it directly.

**Comment by PaddleStroke** [comment]
Got it.
Yes of course please go ahead.

**Comment by wwmayer** [comment]
d119c1c7a5

File: src/Gui/NavigationStyle.cpp (hunk 1 above)
Comment (0penBrain) [reply src/Gui/NavigationStyle.cpp:1483]: Seconds @wwmayer here using Inventor API for code consistency.
Also the 2 'if' can be joigned together with `&&` for clarity.
